### ISSUE TITLE
NPE with BackgroundTupleResult.handleClose in stacktrace

### ISSUE THREAD

**Issue by ansell** [eclipse/rdf4j#293]
The Jenkins server managed to find another NPE with BackgroundTupleResult.handleClose in the stacktrace. This is likely related to BackgroundTupleResult operating in its own thread and needing to be accessed by both cleanup threads and the originating thread.

https://hudson.eclipse.org/rdf4j/job/rdf4j-verify-pr/10/testReport/junit/org.eclipse.rdf4j.repository.sparql/SPARQLStoreConnectionTest/testQueryBaseURI_NONE_/

Similar issues in the past have been reported at:

https://openrdf.atlassian.net/browse/SES-1978

https://openrdf.atlassian.net/browse/SES-1039

https://bitbucket.org/openrdf/sesame/pull-requests/134/update-to-apache-httpcomponents-httpclient/diff#comment-293691

This was discovered solely on the Jenkins server while testing pull request #290 , but as it cannot be replicated locally and that pull request does not modify queryresultio code, it is not likely related to it specifically.

**Comment by ansell** [comment]
Breaking down the stack trace gives the likely cause as being in EofSensorInputStream or ResponseEntityProxy because it allows multiple calls to close, which should be legitimate given the contract of java.io.Closeable.close that says that "If the stream is already closed then invoking this method has no effect."

First line is an NPE where the wrappedStream is accessed without a null-guard, although given it is an input parameter to the method, one could argue that the issue is with EofSensorInputStream for not checking before sending it up the chain.

```
java.lang.NullPointerException: null
	at org.apache.http.impl.execchain.ResponseEntityProxy.streamClosed(ResponseEntityProxy.java:140)
```

The next two calls, include methods that contain a finally to switch the wrappedStream to null (amongst other methods in the class, and the general contract, that sets it to null without synchronisation in any of the locations).

```
	at org.apache.http.conn.EofSensorInputStream.checkClose(EofSensorInputStream.java:228)
	at org.apache.http.conn.EofSensorInputStream.close(EofSensorInputStream.java:174)
```

There is an annotation on EofSensorInputStream that says "@NotThreadSafe", so it would be worthless to report a threading bug on such a class.

Hence, we need to do explicit synchronisation before every call down into Apache HTTPClient, given that they declare internally, and use broken null checking/setting patterns themselves.

**Comment by ansell** [comment]
I have opened an issue and sent a patch upstream for the core issue that triggered the audit here. If it doesn't miss deadlines, the fix may be present in httpclient-4.5.3:

https://issues.apache.org/jira/browse/HTTPCLIENT-1767

**Comment by ansell** [comment]
Fixing this issue included an audit of the cases where the handleClose method implementations did not follow the basic assumption that each layered iteration would consistently (using try-finally) call the handleClose method from any super-classes during execution of its handleClose method.

As part of that audit, there were many cases in the RDF4J codebase where the handleClose and close methods were found not to be null-safe in cases where non-final variables were dereferenced after null-guards. As this was the cause of the underlying issue with the upstream Apache HttpClient library, it seemed reasonable to fix those issues by dereferencing variables consistently before null-guards, and adding null-guards before uses of both mutable variables and nullable final variables.

The audit also revealed many cases where CloseableIteration.close was not being consistently called, resulting in AbstractCloseableIteration.handleClose not being consistently called and a fully consistent try-finally pattern was added for those cases.

In verifying the changes above, and given that we have a Java-7 base version for code patterns (until we abandon the version 1 series), try-with-resources was also used to simplify resource closures where possible. There were also many cases where @Override annotations, that help signal compiler failures after incompatible changes, were not consistently used, which has been allowed for interfaces since Java-6 but has not been updated since creation. There were also some cases where a close method that is compatible with AutoCloseable was present, but the class did not implement AutoCloseable so IDEs and other tools would not automatically find resource closure issues relating to it. In these cases, the relevant classes were modified to implement either AutoCloseable or Closeable and @Override annotations were added to verify the close method signature is maintained in future.

In some cases try-with-resources was not able to be used because the iteration needed to persist beyond the current method. In those cases, a pattern was implemented that closed only when failures occurred, and ensured that in the failure cases all of the resources were released by calling each of their close methods in try-finally blocks within an overall finally block.

There were also some redundant close instance variables in AbstractCloseableIteration subclasses that have now been converted to all use a common method from AbtractCloseableIteration.

In order to consistently trigger the close process to prevent resource leaks, the close-on-exhaust pattern was also implemented for subclasses of AbstractCloseableIteration for hasNext and next when the iteration received enough information to signal that future calls were not going to be successful given the contract for the Iteration interface. 

Although there are some anecdotal arguments that close-on-exhaust may cause separate programming errors to be obscured, resource leaks are a more common occurrence in general and they should be consistently handled wherever possible. Major IDE's and code evaluation tools are able to signal resource closure issues for developers who wish to use them, as CloseableIteration now extends from java.util.AutoCloseable that was introduced in Java-7 and is required for the try-with-resources pattern. 

The only known remaining case where an implementation of CloseableIteration does not actively and fully close its resources is the MutableTupleQueryResult class. Although it is not compatible in general with the CloseableIteration contract, it has a particular purpose that needed to be maintained. It relies on the JVM garbage collector to detect when all in-memory references to it have gone out of scope, and does not keep references to disk or network resources.
